radixdlt-engine/src/main/java/com/radixdlt/atommodel/tokens/scrypt/TokensConstraintScryptV3.java

```diff
@@ -0,0 +1,180 @@
+import com.radixdlt.constraintmachine.EndProcedure;
+import com.radixdlt.constraintmachine.PermissionLevel;
+import com.radixdlt.constraintmachine.ProcedureException;
+import com.radixdlt.constraintmachine.ReducerResult;
+import com.radixdlt.constraintmachine.ReducerState;
+import com.radixdlt.constraintmachine.UpProcedure;
+import com.radixdlt.constraintmachine.VoidReducerState;
+import com.radixdlt.utils.UInt384;
+
+public final class TokensConstraintScryptV3 implements ConstraintScrypt {
+	@Override
+	public void main(Loader os) {
+		registerParticles(os);
+		defineTokenCreation(os);
+		defineMintTransferBurn(os);
+	}
+
+	private void registerParticles(Loader os) {
+		os.particle(
+			TokenResource.class,
+			ParticleDefinition.<TokenResource>builder()
+				.staticValidation(TokenDefinitionUtils::staticCheck)
+				.build()
+		);
+
+		os.particle(
+			TokensInAccount.class,
+			ParticleDefinition.<TokensInAccount>builder()
+				.staticValidation(TokenDefinitionUtils::staticCheck)
+				.build()
+		);
+	}
+
+	private static class NeedFixedTokenSupply implements ReducerState {
+		private final byte[] arg;
+		private final TokenResource tokenResource;
+		private NeedFixedTokenSupply(byte[] arg, TokenResource tokenResource) {
+			this.arg = arg;
+			this.tokenResource = tokenResource;
+		}
+	}
+
+	private void defineTokenCreation(Loader os) {
+		os.procedure(new UpProcedure<>(
+			CMAtomOS.REAddrClaim.class, TokenResource.class,
+			u -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+			(s, u, c, r) -> {
+				if (!u.getAddr().equals(s.getAddr())) {
+					throw new ProcedureException("Addresses don't match");
+				}
+
+				if (u.isMutable()) {
+					return ReducerResult.complete();
+				}
+
+				return ReducerResult.incomplete(new NeedFixedTokenSupply(s.getArg(), u));
+			}
+		));
+
+		os.procedure(new UpProcedure<>(
+			NeedFixedTokenSupply.class, TokensInAccount.class,
+			u -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+			(s, u, c, r) -> {
+				if (!u.getResourceAddr().equals(s.tokenResource.getAddr())) {
+					throw new ProcedureException("Addresses don't match.");
+				}
+
+				if (!u.getAmount().equals(s.tokenResource.getSupply().orElseThrow())) {
+					throw new ProcedureException("Initial supply doesn't match.");
+				}
+
+				return ReducerResult.complete();
+			}
+		));
+	}
+
+	private void defineMintTransferBurn(Loader os) {
+		// Mint
+		os.procedure(new UpProcedure<>(
+			VoidReducerState.class, TokensInAccount.class,
+			u -> {
+				if (u.getResourceAddr().isNativeToken()) {
+					return new Authorization(PermissionLevel.SYSTEM, (r, c) -> { });
+				}
+
+				return new Authorization(PermissionLevel.USER, (r, c) -> {
+					var tokenDef = (TokenResource) r.loadAddr(null, u.getResourceAddr())
+						.orElseThrow(() -> new AuthorizationException("Invalid token address: " + u.getResourceAddr()));
+					tokenDef.verifyMintAuthorization(c.key());
+				});
+			},
+			(s, u, c, r) -> {
+				c.verifyCanDestroyResources();
```

Comment:
Can we rename `verifyCanDestroyResources` to indicate that it controls both allocate and deallocate  resources?